File: src/main/java/com/laconic/pcms/entity/Project.java
package com.laconic.pcms.entity;

import com.laconic.pcms.enums.ProgressStatus;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.Fetch;
import org.hibernate.annotations.FetchMode;
import org.hibernate.annotations.Where;

import java.util.*;

@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@NamedEntityGraph(name = "project.graph", attributeNodes = {
        @NamedAttributeNode("company"),
        @NamedAttributeNode("user"),
        @NamedAttributeNode("stage"),
        @NamedAttributeNode(value = "taskStages", subgraph = "taskStagesGraph"),
        @NamedAttributeNode("space"),
        @NamedAttributeNode(value = "tasks", subgraph = "tasksGraph")
},
        subgraphs = {
                @NamedSubgraph(name = "taskStagesGraph", attributeNodes = {
                        @NamedAttributeNode("user")
                }),
                @NamedSubgraph(name = "tasksGraph", attributeNodes = {
                        @NamedAttributeNode("subTasks"),
                })
        })
@Where(clause = "is_active = true")
public class Project extends BaseEntity<String> {
    @Id
    @GeneratedValue(strategy = GenerationType.TABLE, generator = "id_generator")
    @TableGenerator(name = "id_generator", table = "id_generator", pkColumnName = "id_name", valueColumnName = "next_id", allocationSize = 1)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "company_id")
    private Company company;

    private String color;
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private User user; // responsible

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "stage_id")
    private Stage stage;

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "project_task_stage",
            joinColumns = @JoinColumn(name = "project_id"),
            inverseJoinColumns = @JoinColumn(name = "task_stage_id")
    )
    @Fetch(FetchMode.SUBSELECT)
    private Set<TaskStage> taskStages;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "space_id")
    private Space space;

    private String name;
    private String label;
    private String description;
    private Boolean isRecurringAllowed;
    private String lastUpdateStatus;
    private Long lastUpdateId;
    private Date deadlineDate;
    private Date startDate;
    private String allocatedHours;
    @Builder.Default
    private Double progress = 0.0;
    private String url;
    @Builder.Default
    private Boolean isPrivate = false;

    @OneToMany(mappedBy = "project", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private Set<Task> tasks;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "folder_id")
    private Folder folder;

    @Enumerated(EnumType.STRING)
    ProgressStatus status;
}

File: client/src/types/Project.ts
/**
 * Represents a project with the following properties
 * 
 * @property {number | null} id - The unique identifier for the project. It can be a number or null.
 * @property {number | null} companyId - The identifier of the company to which the project belongs (optional).
 * @property {string | undefined} companyName - The name of the company (optional).
 * @property {number | null} spaceId - The identifier of the space to which the project belongs.
 * @property {string | undefined} spaceName - The name of the space (optional).
 * @property {number | null} folderId - The identifier of the folder to which the project belongs.
 * @property {string | undefined} folderName - The name of the folder (optional).
 * @property {number | null} userId - The identifier of the user associated with the project (optional).
 * @property {string | undefined} userName - The name of the user (optional).
 * @property {number} stageId - The identifier of the stage to which the project belongs.
 * @property {string | undefined} stageName - The name of the stage (optional).
 * @property {string} color - The color associated with the project.
 * @property {string} name - The name of the project.
 * @property {string} url - The URL of the project.
 * @property {string | undefined} label - A label associated with the project (optional).
 * @property {string | undefined} description - A description of the project (optional).
 * @property {number | undefined} progress - The progress of the project (optional).
 * @property {Date | string | null} start - The start date of the project (optional).
 * @property {Date | string | null} end - The end date of the project (optional).
 * @property {string | undefined} lastUpdateStatus - The status of the last update (optional).
 * @property {string | undefined} allocatedHours - The allocated hours for the project (optional).
 * @property {TaskStage[]} taskStages - An array of task stages associated with the project.
 * @property {Task[] | undefined} tasks - An array of tasks associated with the project (optional).
 * @property {boolean | undefined} recurringAllowed - Indicates whether recurring tasks are allowed (optional).
 * @property {boolean} isPrivate - Indicates whether the project is private or not.
 * @property {boolean | undefined} isFavorite - Indicates whether the project is marked as a favorite (optional).
 * @property {boolean | undefined} isActive - Indicates whether the project is active or not (optional).
 */
type Project = {
    id: number | null;
    companyId?: number | null;
    companyName?: string;
    spaceId: number | null;
    spaceName?: string;
    folderId: number | null;
    folderName?: string;
    userId: number | null;
    userName?: string;
    stageId: number;
    stageName?: string;
    color: string;
    name: string;
    url: string;
    label?: string;
    description?: string;
    progress?: number;
    start?: Date | string | null;
    end?: Date | string | null;
    lastUpdateStatus?: string;
    allocatedHours?: string;
    taskStages: TaskStage[];
    tasks?: Task[];
    recurringAllowed?: boolean;
    isPrivate: boolean;
    isFavorite?: boolean;
    isActive?: boolean;
}
